backend/src/apiserver/storage/pipeline_store_kubernetes.go

```diff
@@ -0,0 +1,502 @@
+			if err != nil {
+				return nil, 0, "", err
+			}
+
+			if opts.Filter == nil {
+				pipelineVersions = append(pipelineVersions, pipelineVersion)
+			} else if opts.Filter.FilterK8sPipelineVersions(k8sPipelineVersion) {
+				pipelineVersions = append(pipelineVersions, pipelineVersion)
+			}
+		}
+	}
+
+	// Because controller-client does not have sorting, use this function to sort by fields.
+	sort.Slice(pipelineVersions, func(i, j int) bool {
+		if strings.Compare(opts.SortByFieldName, "CreatedAtInSec") == 0 {
+			return pipelineVersions[i].CreatedAtInSec > pipelineVersions[j].CreatedAtInSec
+		} else if strings.Compare(opts.SortByFieldName, "Name") == 0 {
+			return pipelineVersions[i].Name > pipelineVersions[j].Name
+		}
+		return false
+	})
+	if !opts.IsDesc {
+		slices.Reverse(pipelineVersions)
+	}
+
+	// If there's pagination, find the index of the first element of the next page
+	nextPage := 0
+	for p, pipelineVersion := range pipelineVersions {
+		if strings.Compare(fmt.Sprint(opts.KeyFieldValue), pipelineVersion.UUID) == 0 {
+			nextPage = p
+		}
+	}
+
+	// Split results on multiple pages if needed
+	lastIndex := nextPage + opts.PageSize
+	npt := ""
+	if len(pipelineVersions) <= opts.PageSize {
+		return pipelineVersions, len(pipelineVersions), "", nil
+	} else if lastIndex >= len(pipelineVersions) {
+		lastIndex = len(pipelineVersions)
+		npt, err = opts.NextPageToken(pipelineVersions[lastIndex-1])
+	} else {
+		npt, err = opts.NextPageToken(pipelineVersions[lastIndex])
+	}
+	return pipelineVersions[nextPage:lastIndex], len(pipelineVersions), npt, err
+}
+
+func (k *PipelineStoreKubernetes) DeletePipelineVersion(pipelineVersionId string) error {
+	k8sPipelineVersion, err := k.getK8sPipelineVersion(pipelineVersionId)
+	if err != nil {
+		if strings.Contains(err.Error(), "ResourceNotFoundError:") {
+			return nil
+		}
+	}
+
+	err = k.client.Delete(context.TODO(), k8sPipelineVersion)
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return util.NewInternalServerError(err, "Failed to delete the pipeline version")
+	}
+
+	return nil
+}
+
+func (k *PipelineStoreKubernetes) getK8sPipeline(pipelineId string) (*v2beta1.Pipeline, error) {
+	// TODO: Consider making the ID namespace/name format to make querying efficient
+	pipelines := v2beta1.PipelineList{}
+
+	listOptions := []ctrlclient.ListOption{ctrlclient.UnsafeDisableDeepCopy}
+
+	if !common.IsMultiUserMode() && common.GetPodNamespace() != "" {
+		listOptions = append(listOptions, ctrlclient.InNamespace(common.GetPodNamespace()))
+	}
+
+	// Be careful, the deep copy is disabled here to reduce memory allocations
+	err := k.client.List(context.TODO(), &pipelines, listOptions...)
+	if err != nil {
+		return nil, util.NewInternalServerError(err, "Failed to find the pipeline")
+	}
+
+	for _, k8sPipeline := range pipelines.Items {
+		if string(k8sPipeline.UID) == pipelineId {
+			return &k8sPipeline, nil
+		}
+	}
+
+	return nil, util.NewResourceNotFoundError("Pipeline", pipelineId)
```

Comment:
have you considered storing the UID as a label and just doing a listing option as fieldselector?

Reply:
Will be implemented when webhooks PR is merged

Reply:
You would need a controller to do that since I'm fairly certain that the UID wouldn't be available for the mutating webhook. Filtering pipeline versions by pipeline ID will be possible.

Reply:
thinking out loud, you could store the UID after the first .get() to make future calls more efficient, but we don't have to block the pr on this

Reply:
You can't without a controller because the pipeline version could be created through Kubernetes directly.

Reply:
I meant after creation, and for the label, but not strongly opinionated on this right now